[SamplePGO] Support -salvage-stale-profile without probes too

amharc opened the pull request:
Currently -salvage-stale-profile is a no-op if the profile is not probe-based. We observed that it can help for regular, non-probe- based profiles too: some of our internal benchmarks show 0.2-0.3% QPS improvement.

There seems to be no good reason to limit this flag to only work for probe-based profiles.

amharc committed:
[SamplePGO] Support -salvage-stale-profile without probes too

Currently -salvage-stale-profile is a no-op if the profile is not
probe-based. We observed that it can help for regular, non-probe-
based profiles too: some of our internal benchmarks show 0.2-0.3%
QPS improvement.

There seems to be no good reason to limit this flag to only work
for probe-based profiles.

WenleiHe commented:
> There seems to be no good reason to limit this flag to only work for probe-based profiles.

This should help AutoFDO too. We didn't turn it on because we didn't test it throughly for AutoFDO.

Also note that the matching algorithm may not be a no-op for perfectly matched profile. Since we don't have checksum for AutoFDO, we can only apply matching to all functions, which may have undesired side effects for good/matched profiles.

wlei-llvm commented:
FWIW, there is a TODO of LTO time matching for AutoFDO path, see: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/IPO/SampleProfile.cpp#L2239, it doesn't extract the non-callsite line-location anchor,  so the regular instruction still uses the old mapping which may lead to inaccurate block count. At that time, we only compute the staleness stats for AutoFDO path, so I didn't add the support to save some build-time.

wlei-llvm commented on llvm/test/Transforms/SampleProfile/non-probe-stale-profile-matching.ll:
```diff
@@ -0,0 +1,231 @@
+!llvm.dbg.cu = !{!2, !7}
+!llvm.module.flags = !{!9, !10, !11, !12, !13, !14, !15, !16, !17}
+!llvm.ident = !{!46, !46}
+!llvm.pseudo_probe_desc = !{!47, !48}
```
Looks the IR is still built with pseudo-probe, but the profile is non-probe, which is a mismatch. Though it can work, but would it be better to use a non-probe IR? It might be confusing in the future if someone changes pseudo-probe features but it would break a non-probe-xxx.ll test.

amharc replied:
Good catch. I was initially lazy and just copied one of the files for pseudo-probe-based stale profile matching and removed pseudoprobe instructions :) I regenerated the IR (and the stale profile) in non-psuedo-probe mode now.

wlei-llvm commented on llvm/lib/Transforms/IPO/SampleProfile.cpp:
```diff
@@ -2411,10 +2411,10 @@ void SampleProfileMatcher::runOnFunction(const Function &F) {
-  // support for pseudo-probe.
-  if (SalvageStaleProfile && FunctionSamples::ProfileIsProbeBased &&
-      !ProbeManager->profileIsValid(F, *FSFlattened)) {
+  // Run profile matching for checksum mismatched profile. For probe-based
```
Nit: Rephrase or remove this "Run profile matching for checksum mismatched profile", since non-probe doesn't have checksum.

amharc replied:
Done.

amharc commented:
> FWIW, there is a TODO of LTO time matching for AutoFDO path, see: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/IPO/SampleProfile.cpp#L2239, it doesn't extract the non-callsite line-location anchor, so the regular instruction still uses the old mapping which may lead to inaccurate block count.

Yes, I'm aware of this TODO and I plan to work on it afterwards. Properly matching just callsite line-locations helps a lot on itself too: it affects the inlining behaviour of the sample profile loader. If `Foo` calls `Bar` calls `Baz`, adding new lines in `Foo` above the will no longer cause the whole profile for `Baz` inlined into `Bar` inlined into `Foo` to be dropped with this change (and with `-mllvm=-salvage-stale-profile=true`). Real life functions tend to be short, so the performance loss from non-call lines in `Foo` having wrong profile weights shouldn't matter as much as the performance loss from the total lack of past inlining information (and profiles contextualised to the inlined instances).

wlei-llvm reviewed (approved):
LGTM, thanks.  Just a friendly heads-up, there is a bug related to callee profile matches, and I'm working on the fix(see #85715), it works for checksum(probe-based) profile. But as Wenlei pointed out,  there is no checksum thing for AutoFDO, so even the user function code is not changed, i,e. the profile is not stale, all functions(after the fix, all callees) will run the "stale" profile matching, and it might not be a no-op(e.g. cold or optimized-out callsite might not show in the profile, hopefully cold callsite are insignificant, but I'm not sure) and might cause negative effect to this perfectly matched case.

amharc commented:
Thanks for the review! I rebased the changes (there were some merge conflicts) and squashed them into one commit. I don't have write access to LLVM (and in the light of what happened last week I'm not sure if we all want to have even more people with write access to critical infrastructure...), so I'd like to ask someone to merge the PR.
